### ISSUE TITLE
Cache with mongoose ObjectId will fail to find the keyFieldName in cache

### ISSUE THREAD

**Issue by axmad22** [feathers-plus/feathers-hooks-common#424]
First just wanted to thank you for this awesome tool, and wanted to inform that when using the cache hook with feathers mongoose as a service in the after method it will return a document with the _id field of type ObjectId not a string, thus will fail to find it when searching the cache index. 

Example of a way am calling a service and returning a document
``` js
class BlogModel {
    async get(id, params) {
       return app.getService('posts').get(id, {query: {$populate: 'owner'}});
     }
}
module.exports = {
  before: {
    all: cache(cacheMap, '_id')
  },
  after: {
    all: cache(cacheMap, '_id')
  }
};
```

The fix am using, or should I have used a transformation hook before the caching

``` js
'use strict';

const common = require('feathers-hooks-common');
const mongoose = require("mongoose");

module.exports = function (cacheMap, keyFieldName, options = {}) {
  const clone = options.clone || defaultClone;
  return context => {
    keyFieldName = keyFieldName || (context.service || {}).id; // Will be undefined on client

    let items = common.getItems(context);
    items = Array.isArray(items) ? items : [items];

    if (context.type === 'after') {
      if (context.method === 'remove') {
        return;
      }

      items.forEach(item => {
        const idName = getIdName(keyFieldName, item);
        // cacheMap.set(item[idName], clone(item));
        cacheMap.set(objectIdToString(item[idName]), clone(item));
      });
      return;
    }

    switch (context.method) {
      case 'find': // fall through
      case 'create':
        return;
      case 'get': {
        const value = cacheMap.get(context.id);
        if (value) {
          context.result = value;
        }
        return context;
      }
      default: // update, patch, remove
        if (context.id) {
          cacheMap.delete(context.id);
          return;
        }

        items.forEach(item => {
          const idName = getIdName(keyFieldName, item);
          // cacheMap.delete(item[idName]);
          cacheMap.delete(objectIdToString(item[idName]));
        });
    }
  };
};

function getIdName (keyFieldName, item) {
  if (keyFieldName) return keyFieldName;
  return ('_id' in item) ? '_id' : 'id';
}

function defaultClone (obj) {
  return JSON.parse(JSON.stringify(obj));
}

function objectIdToString (id) {
 return (id instanceof mongoose.Types.ObjectId) ? id.toString() : id;
}
```

**Comment by eddyystop** [comment]
Could you submit a PR for this?

We would need a new, optional `option.convertrKey`. It would be an optional function like your objectIdToString.

Its executed, if it exists, in after hooks. Does it make sense to also execute it in before hooks?

That way the hook doesn't require Mongoose/MongoDB.

Thanks!

@marshallswain @Mattchewone could you please also comment on the above?

**Comment by marshallswain** [comment]
Yep.  There’s a mismatch when one version of the _id is an object and the other is the stringified version of that object.

**Comment by axmad22** [comment]
Am sorry I don't use any version controls to create a PR. I wanted to ask your opinion after discovering the cache is it a good idea to use it on the User service or will it effect authentication, socket.io  and other Services. Thanks

**Comment by eddyystop** [comment]
The users service would be one of the better places to use `cache` if possible.

There are several complications:
- Things get complicated when hooks make service calls themselves. `authenticate` and `restrictToUser` both call the users service *if its an external call*.
- Setting `context.result`, as `cache` does, does not prevent the following hooks from running. It just prevents the DB call once all the before hooks have run.
- Most before hooks do not check if `context.result` is set. They assume they have to process `context.data`. Such hooks would ignore the record found by `cache`, process `context.data` needlessly as `context.data` found be ignored as the DB call is not made.

Can you use `cache` on the users service depends on the circumstances.

The good news is `restrictToUser` and `authenticate` return immediately if the call has been made by the server. So if your hooks are:
```js
before: { get: [ cache(), restrictToUser(), authenticate() ] },
after: { get: [ cache(), discard('password') }
```
The following hooks execute:
- on the external call
  - cache before
  - restrictToUser, which makes a get call
    - cache before
    - no DB call
    - cache after
  - authenticate, which makes a get call
    - cache before
    - no DB call
    - cache after
  - no DB call
  - cache after
  - discard

Which looks like it *should* work. Let me know your results should you decide to try it.

**Comment by eddyystop** [comment]
Fixed in https://github.com/feathers-plus/feathers-hooks-common/pull/427

Published as v4.16.2.

Please reply if this handles your user case. Please also review the docs at https://feathers-plus.github.io/v1/feathers-hooks-common/#cache

**Comment by axmad22** [comment]
Yes its working now, also am testing it on the Users service I had to make sure in every place in code where I test for the users ID or any ID to do it using a helper function,  since it could be either a string or an ObjectId depending if it came from the cacheMap or DB now.  Its nice especially with web-socket it stopped hitting the DB, and that kind of feels wrong with the users case am not sure if its safe but I will keep on testing and see what happens.
thanks for the help

// helper function
``` js
example testing in code
if (compareIds(context.params.user._id, post.createdBy)) {
}

const compareIds = (idOne, idTwo) => {
  if (idOne && idTwo) {
    let tmpOne = (idOne instanceof mongoose.Types.ObjectId) ? idOne.toString() : idOne;
    let tmpTwo = (idTwo instanceof mongoose.Types.ObjectId) ? idTwo.toString() : idTwo;

    return tmpOne === tmpTwo;
  }
  return false;
};
```


Reviewed thedocs and the makeCacheKey in the docs should have been

``` js
  // it should be key and not id
  const makeCacheKey = key => key instanceof mongoose.Types.ObjectId ? key .toString() : key ;
  const makeCacheKey = key => id instanceof mongoose.Types.ObjectId ? id.toString() : id;
```
